test/puppetlabs/http/client/metrics_test.clj

```diff
@@ -0,0 +1,481 @@
+                    client-metrics-data (common/get-client-metrics-data client)
+                    base-metric-id (str "puppetlabs.http-client.experimental.with-url.http://localhost:" port "/hello")
+                    bytes-read-id (str base-metric-id ".bytes-read")
+                    bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                       (set (keys client-metrics))
+                       (set (keys client-metrics-data))))
+                (is (every? #(instance? Timer %) (vals client-metrics)))
+                (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                  (is (= {:count 1 :metric-id bytes-read-id}
+                         (select-keys bytes-read-data [:metric-id :count])))
+                  (is (<= 1000 (:mean bytes-read-data)))
+                  (is (<= 1000 (:aggregate bytes-read-data))))))))))
+     (testing "metrics work for a failed request"
+       (try
+         (testwebserver/with-test-webserver-and-config
+          (unbuffered-test/blocking-handler data) port {:shutdown-timeout-seconds 1}
+          (let [metric-registry (MetricRegistry.)]
+            (with-open [client (async/create-client {:connect-timeout-milliseconds 100
+                                                     :socket-timeout-milliseconds 200}
+                                                    metric-registry)]
+              (let [response @(common/get client (str "http://localhost:" port "/hello") opts)
+                    {:keys [body error]} response]
+                (is (nil? error))
+                ;; Consume the body to get the exception
+                (is (thrown? SocketTimeoutException (slurp body))))
+              (let [client-metrics (common/get-client-metrics client)
+                    client-metrics-data (common/get-client-metrics-data client)
+                    base-metric-id (str "puppetlabs.http-client.experimental.with-url.http://localhost:" port "/hello")
+                    bytes-read-id (str base-metric-id ".bytes-read")
+                    bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                       (set (keys client-metrics))
+                       (set (keys client-metrics-data))))
+                (is (every? #(instance? Timer %) (vals client-metrics)))
+                (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                  (is (= {:count 1 :metric-id bytes-read-id}
+                         (select-keys bytes-read-data [:metric-id :count])))
+                  (is (<= 1 (:mean bytes-read-data)))
+                  (is (<= 1 (:aggregate bytes-read-data))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
+           ))))))
```

Comment:
Same question here as for the corresponding Java test on when we'd expect it to be normal for this exception to be thrown.